ProxyFactory createProxyName generates a NPE when not in an enclosing type

* Helidon Version: 2.5.0
* Helidon MP
* JDK version: Java(TM) SE Runtime Environment (build 17.0.2+8-LTS-86)
* OS: Ubuntu 21.10
----------

## Problem Description
The use case is that I'm trying to use a CDI based decorator to invoke a MapStruct mapper.
A `@Decorator` applied on a top-level class yields a NPE in the `createProxyName` call of the augmented `ProxyFactory` class 

```
➜  target git:(master) ✗ java -jar helidon-mapstruct-mp.jar
2022.04.13 13:21:28 INFO io.helidon.common.LogConfig Thread[main,5,main]: Logging at initialization configured using classpath: /logging.properties
Exception in thread "main" java.lang.ExceptionInInitializerError
        at io.helidon.microprofile.cdi.ContainerInstanceHolder.fromBuildTimeInitializer(ContainerInstanceHolder.java:58)
        at io.helidon.microprofile.cdi.ContainerInstanceHolder.get(ContainerInstanceHolder.java:50)
        at io.helidon.microprofile.cdi.Main.<clinit>(Main.java:55)
Caused by: java.lang.NullPointerException
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
        at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:564)
        at java.base/java.util.concurrent.ForkJoinTask.reportExecutionException(ForkJoinTask.java:604)
        at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:981)
        at org.jboss.weld.executor.AbstractExecutorServices.checkForExceptions(AbstractExecutorServices.java:73)
        at org.jboss.weld.executor.AbstractExecutorServices.invokeAllAndCheckForExceptions(AbstractExecutorServices.java:59)
        at org.jboss.weld.executor.AbstractExecutorServices.invokeAllAndCheckForExceptions(AbstractExecutorServices.java:67)
        at org.jboss.weld.bootstrap.ConcurrentBeanDeployer.createClassBeans(ConcurrentBeanDeployer.java:65)
        at org.jboss.weld.bootstrap.BeanDeployment.createBeans(BeanDeployment.java:256)
        at org.jboss.weld.bootstrap.WeldStartup.deployBeans(WeldStartup.java:438)
        at org.jboss.weld.bootstrap.WeldBootstrap.deployBeans(WeldBootstrap.java:86)
        at io.helidon.microprofile.cdi.HelidonContainerImpl.init(HelidonContainerImpl.java:215)
        at io.helidon.common.context.Contexts.runInContext(Contexts.java:137)
        at io.helidon.microprofile.cdi.HelidonContainerImpl.initInContext(HelidonContainerImpl.java:136)
        at io.helidon.microprofile.cdi.HelidonContainerImpl.create(HelidonContainerImpl.java:128)
        at io.helidon.microprofile.cdi.BuildTimeInitializer.createContainer(BuildTimeInitializer.java:51)
        at io.helidon.microprofile.cdi.BuildTimeInitializer.<clinit>(BuildTimeInitializer.java:31)
        ... 3 more
Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Class.getName()" because "superInterface" is null
        at org.jboss.weld.bean.proxy.ProxyFactory.createProxyName(ProxyFactory.java:340)
        at org.jboss.weld.bean.proxy.ProxyFactory.getProxyName(ProxyFactory.java:280)
        at org.jboss.weld.bean.proxy.ProxyFactory.<init>(ProxyFactory.java:191)
        at org.jboss.weld.bean.proxy.ProxyFactory.<init>(ProxyFactory.java:187)
        at org.jboss.weld.bean.proxy.DecoratorProxyFactory.<init>(DecoratorProxyFactory.java:68)
        at org.jboss.weld.injection.producer.DecoratorInjectionTarget$1.createEnhancedSubclass(DecoratorInjectionTarget.java:78)
        at org.jboss.weld.injection.producer.SubclassedComponentInstantiator.initEnhancedSubclass(SubclassedComponentInstantiator.java:86)
        at org.jboss.weld.injection.producer.SubclassedComponentInstantiator.<init>(SubclassedComponentInstantiator.java:79)
        at org.jboss.weld.injection.producer.DecoratorInjectionTarget$1.<init>(DecoratorInjectionTarget.java:75)
        at org.jboss.weld.injection.producer.DecoratorInjectionTarget.initInstantiator(DecoratorInjectionTarget.java:75)
        at org.jboss.weld.injection.producer.BasicInjectionTarget.<init>(BasicInjectionTarget.java:94)
        at org.jboss.weld.injection.producer.BasicInjectionTarget.<init>(BasicInjectionTarget.java:78)
        at org.jboss.weld.injection.producer.BeanInjectionTarget.<init>(BeanInjectionTarget.java:58)
        at org.jboss.weld.injection.producer.BeanInjectionTarget.<init>(BeanInjectionTarget.java:63)
        at org.jboss.weld.injection.producer.DecoratorInjectionTarget.<init>(DecoratorInjectionTarget.java:58)
        at org.jboss.weld.manager.InjectionTargetFactoryImpl.chooseInjectionTarget(InjectionTargetFactoryImpl.java:124)
        at org.jboss.weld.manager.InjectionTargetFactoryImpl.createInjectionTarget(InjectionTargetFactoryImpl.java:88)
        at org.jboss.weld.bean.ManagedBean.<init>(ManagedBean.java:102)
        at org.jboss.weld.bean.DecoratorImpl.<init>(DecoratorImpl.java:61)
        at org.jboss.weld.bean.DecoratorImpl.of(DecoratorImpl.java:50)
        at org.jboss.weld.bootstrap.AbstractBeanDeployer.createDecorator(AbstractBeanDeployer.java:293)
        at org.jboss.weld.bootstrap.BeanDeployer.createClassBean(BeanDeployer.java:214)
        at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$2.doWork(ConcurrentBeanDeployer.java:68)
        at org.jboss.weld.bootstrap.ConcurrentBeanDeployer$2.doWork(ConcurrentBeanDeployer.java:65)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:62)
        at org.jboss.weld.executor.IterativeWorkerTaskFactory$1.call(IterativeWorkerTaskFactory.java:55)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedInterruptibleCallable.exec(ForkJoinTask.java:1461)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

## Steps to reproduce
I have a reproducer available. Check [here](https://github.com/jvissers/helidon-mapstruct-mp).

--- a/microprofile/weld/weld-core-impl/src/main/java/org/jboss/weld/bean/proxy/ProxyFactory.java
+++ b/microprofile/weld/weld-core-impl/src/main/java/org/jboss/weld/bean/proxy/ProxyFactory.java
@@ -273,13 +273,16 @@ public class ProxyFactory<T> implements PrivilegedAction<T> {
         final String className;
 
         if (typeInfo.getSuperClass() == Object.class) {
-            final StringBuilder name = new StringBuilder();
-
             // for classes that do not have an enclosing class, we want the super interface to be first
             if (proxiedBeanType.getEnclosingClass() == null) {
+                if (typeInfo.getSuperInterface() == null) {
+                    // abstract decorators fall into this category, let's use the type name
+                    return proxiedBeanType.getName() + PROXY_SUFFIX;
+                }
                 return createProxyName(typeInfo) + PROXY_SUFFIX;
             } else {
                 //interface only bean.
+                final StringBuilder name = new StringBuilder();
                 className = createCompoundProxyName(contextId, bean, typeInfo, name) + PROXY_SUFFIX;
             }
         } else {

--- a/tests/integration/mp-gh-4123/pom.xml
+++ b/tests/integration/mp-gh-4123/pom.xml
@@ -1,0 +1,54 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2022 Oracle and/or its affiliates.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>io.helidon.tests.integration</groupId>
+        <artifactId>helidon-tests-integration</artifactId>
+        <version>2.5.1-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>helidon-tests-integration-mp-gh-4123</artifactId>
+    <name>Helidon Tests Integration MP GH 4123</name>
+    <description>Reproducer for Github issue #4123 - abstract Decorator class fails with NPE</description>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.helidon.microprofile.cdi</groupId>
+            <artifactId>helidon-microprofile-cdi</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.helidon.microprofile.tests</groupId>
+            <artifactId>helidon-microprofile-tests-junit5</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+</project>

--- a/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/ContractBean.java
+++ b/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/ContractBean.java
@@ -1,0 +1,27 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh4123;
+
+import javax.enterprise.context.ApplicationScoped;
+
+@ApplicationScoped
+public class ContractBean implements SomeContract {
+    @Override
+    public String message() {
+        return "fromBean";
+    }
+}

--- a/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/ContractDecorator.java
+++ b/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/ContractDecorator.java
@@ -1,0 +1,39 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh4123;
+
+import javax.decorator.Decorator;
+import javax.decorator.Delegate;
+import javax.inject.Inject;
+
+@Decorator
+public abstract class ContractDecorator implements SomeContract {
+    private final SomeContract delegate;
+
+    @Inject
+    public ContractDecorator(
+            @Delegate
+            SomeContract delegate) {
+
+        this.delegate = delegate;
+    }
+
+    @Override
+    public String message() {
+        return "Decorated " + delegate.message();
+    }
+}

--- a/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/SomeContract.java
+++ b/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/SomeContract.java
@@ -1,0 +1,21 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh4123;
+
+public interface SomeContract {
+    String message();
+}

--- a/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/UsingBean.java
+++ b/tests/integration/mp-gh-4123/src/main/java/io/helidon/tests/integration/gh4123/UsingBean.java
@@ -1,0 +1,34 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh4123;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.inject.Inject;
+
+@ApplicationScoped
+public class UsingBean {
+    private final SomeContract contract;
+
+    @Inject
+    public UsingBean(SomeContract contract) {
+        this.contract = contract;
+    }
+
+    public String message() {
+        return contract.message();
+    }
+}

--- a/tests/integration/mp-gh-4123/src/main/resources/META-INF/beans.xml
+++ b/tests/integration/mp-gh-4123/src/main/resources/META-INF/beans.xml
@@ -1,0 +1,29 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+
+    Copyright (c) 2022 Oracle and/or its affiliates.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+
+<beans xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee
+                           http://xmlns.jcp.org/xml/ns/javaee/beans_2_0.xsd"
+        bean-discovery-mode="annotated">
+
+    <decorators>
+        <class>io.helidon.tests.integration.gh4123.ContractDecorator</class>
+    </decorators>
+</beans>

--- a/tests/integration/mp-gh-4123/src/test/java/io/helidon/tests/integration/gh4123/Gh4123Test.java
+++ b/tests/integration/mp-gh-4123/src/test/java/io/helidon/tests/integration/gh4123/Gh4123Test.java
@@ -1,0 +1,43 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.tests.integration.gh4123;
+
+import javax.inject.Inject;
+
+import io.helidon.microprofile.tests.junit5.HelidonTest;
+
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+@HelidonTest
+class Gh4123Test {
+    private final SomeContract contract;
+
+    @Inject
+    Gh4123Test(SomeContract contract) {
+        this.contract = contract;
+    }
+
+    @Test
+    void testDecorator() {
+        String value = contract.message();
+
+        assertThat(value, is("Decorated fromBean"));
+    }
+}

--- a/tests/integration/pom.xml
+++ b/tests/integration/pom.xml
@@ -48,6 +48,7 @@
         <module>mp-gh-2461</module>
         <module>mp-gh-3246</module>
         <module>mp-gh-3974</module>
+        <module>mp-gh-4123</module>
         <module>kafka</module>
         <module>jms</module>
         <module>config</module>

